### PR TITLE
feat: use aggregated perf scope to log perf statistics per phase

### PR DESCRIPTION
**Pull request by herohua**
###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/4988)

### CODE REVIEW

File: src/Microsoft.DocAsCode.Build.Engine/LinkPhaseHandler.cs [hunk 1]
```diff
@@ -221,14 +224,14 @@ private InternalManifestItem GetManifestItem(FileModel model, SaveResult result)
...
-        private static void Postbuild(HostService hostService)
+        private static void Postbuild(HostService hostService, AggregatedPerformanceScope aggregatedPerformanceScope)
         {
             BuildPhaseUtility.RunBuildSteps(
                 hostService.Processor.BuildSteps,
                 buildStep =>
                 {
                     Logger.LogVerbose($"Processor {hostService.Processor.Name}, step {buildStep.Name}: Postbuilding...");
-                    using (new LoggerPhaseScope(buildStep.Name, LogLevel.Verbose))
+                    using (new LoggerPhaseScope(buildStep.Name, LogLevel.Verbose, aggregatedPerformanceScope))
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Build.Engine/LinkPhaseHandler.cs]: >, aggregatedPerformanceScope [](start = 80, length = 28)

For PostBuild, this LoggerPhaseScope already contains all the files. You can see the prameter below is `hostService.Models`. We do not need aggregate here.

Same for PreBuild.

File: src/Microsoft.DocAsCode.Common/Loggers/AggregatedPerformanceScope.cs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+            {
+                var aggregatedPerformance = aggregatedPerformanceByPhase.Value.Value;
+                Logger.Log(_logLevel, $"Phase '{aggregatedPerformanceByPhase.Key}' runs {aggregatedPerformance.Occurrence} times with average time of {aggregatedPerformance.TotalTimeInMilliseconds / aggregatedPerformance.Occurrence} milliseconds.");
+            }
+        }
+
+        private class AggregatedPerformance
+        {
+            private long _occurrence = 0;
+            private double _totalTimeInMilliseconds = 0.0;
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Common/Loggers/AggregatedPerformanceScope.cs]: >_totalTimeInMilliseconds [](start = 27, length = 24)

volatile

File: src/Microsoft.DocAsCode.Common/Loggers/AggregatedPerformanceScope.cs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+            public long Occurrence => _occurrence;
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Common/Loggers/AggregatedPerformanceScope.cs]: >public long Occurrence => _occurrence; [](start = 12, length = 38)

Could auto property {get; private set} work?

File: src/Microsoft.DocAsCode.Common/Loggers/AggregatedPerformanceScope.cs [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+            aggregatedPerformanceByPhase.Log(elapsedTime.TotalMilliseconds);
+        }
+
+        public void Dispose()
+        {
+            foreach (var aggregatedPerformanceByPhase in _aggregatedPerformanceByPhase.OrderBy(kvp => kvp.Key))
+            {
+                var aggregatedPerformance = aggregatedPerformanceByPhase.Value.Value;
+                Logger.Log(_logLevel, $"Phase '{aggregatedPerformanceByPhase.Key}' runs {aggregatedPerformance.Occurrence} times with average time of {aggregatedPerformance.TotalTimeInMilliseconds / aggregatedPerformance.Occurrence} milliseconds.");
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Common/Loggers/AggregatedPerformanceScope.cs]: >Occurrence [](start = 221, length = 10)

Although current code will not trigger, better to add a zero divider check. Can encapsulate the logic in `AggregatedPerformance.AverageTime`.

File: src/Microsoft.DocAsCode.Common/Loggers/AggregatedPerformanceScope.cs (hunk 2 above)
Comment (herohua) [reply src/Microsoft.DocAsCode.Common/Loggers/AggregatedPerformanceScope.cs]: `double` cannot be `volatile`

File: src/Microsoft.DocAsCode.Common/Loggers/AggregatedPerformanceScope.cs (hunk 3 above)
Comment (herohua) [reply src/Microsoft.DocAsCode.Common/Loggers/AggregatedPerformanceScope.cs]: Nope. `Interlocked` cannot work with properties.

**Pull request merged**